pkg/docker/util.go

```diff
@@ -187,46 +186,27 @@ func ParseImageReference(spec string) (ImageReference, error) {
 	var ref ImageReference
 
 	// TODO replace with docker version once docker/docker PR11109 is merged upstream
-	stream, tag, id := parseRepositoryTag(spec)
+	ref.Name, ref.Tag, ref.ID = parseRepositoryTag(spec)
 
-	repoParts := strings.Split(stream, "/")
-	switch len(repoParts) {
+	if len(ref.Name) == 0 {
+		return ImageReference{}, fmt.Errorf("the docker pull spec %q must not be empty", spec)
+	}
+
+	nameParts := strings.SplitN(ref.Name, "/", 3)
+	switch len(nameParts) {
 	case 2:
-		if strings.Contains(repoParts[0], ":") {
-			// registry/name
-			ref.Registry = repoParts[0]
-			ref.Namespace = "library"
-			ref.Name = repoParts[1]
-			ref.Tag = tag
-			ref.ID = id
-			return ref, nil
+		if strings.ContainsAny(nameParts[0], ":.") {
+			// registry/name, we suppose
+			ref.Registry = nameParts[0]
+			ref.Name = "library/" + nameParts[1]
 		}
-		// namespace/name
-		ref.Namespace = repoParts[0]
-		ref.Name = repoParts[1]
-		ref.Tag = tag
-		ref.ID = id
-		return ref, nil
 	case 3:
-		// registry/namespace/name
-		ref.Registry = repoParts[0]
-		ref.Namespace = repoParts[1]
-		ref.Name = repoParts[2]
-		ref.Tag = tag
-		ref.ID = id
-		return ref, nil
-	case 1:
-		// name
-		if len(repoParts[0]) == 0 {
-			return ref, fmt.Errorf("the docker pull spec %q must be two or three segments separated by slashes", spec)
-		}
-		ref.Name = repoParts[0]
-		ref.Tag = tag
-		ref.ID = id
-		return ref, nil
-	default:
-		return ref, fmt.Errorf("the docker pull spec %q must be two or three segments separated by slashes", spec)
+		// registry(/dir)+/name, we suppose
+		ref.Registry = nameParts[0]
```

Comment:
how do we safely make this assumption?  should we check it for : and . like above?
(are : or . valid characters in a docker image path segment?)